modules/effects/arrow-effects-data/src/test/kotlin/arrow/effects/EffectsSuspendDSLTests.kt

```diff
@@ -0,0 +1,373 @@
+      msg.get() shouldBe const
+    }
+
+    "startFiber" {
+      val const = 1
+      fxTest {
+        fx {
+          val fiber = Dispatchers.Default.startFiber { const }
+          val (n) = fiber.join()
+          n
+        }
+      } shouldBe const
+    }
+
+    "racePair" {
+      fxTest {
+        fx {
+          val race1 = Dispatchers.Default.racePair({ 1 }, { Thread.sleep(100); 2 })
+          val race2 = Dispatchers.Default.racePair({ Thread.sleep(100); 1 }, { 2 })
+          race1.fold({ true }, { false }) &&
+            race2.fold({ false }, { true })
+        }
+      } shouldBe true
+    }
+
+    "parallel" {
+      fxTest {
+        fx {
+          val currentThread = effect { Thread.currentThread().name }
+          val (ta: String, tb: String) = Dispatchers.Default.parallel(
+            { Thread.currentThread().name },
+            { Thread.currentThread().name }
+          )
+          ta != currentThread && tb != currentThread
+        }
+      } shouldBe true
+    }
+
+    "Either component1() syntax success" {
+      fxTest {
+        fx {
+          val (r) = 1.right()
+          r
+        }
+      } shouldBe 1
+    }
+
+    "Either component1() syntax failure" {
+      shouldThrow<TestError> {
+        fxTest {
+          fx {
+            val (r) = TestError.left()
+            r
+          }
+        }
+      }
+    }
+
+    "List.flatTraverse syntax" {
+      fxTest {
+        fx {
+          val result = listOf(
+            suspend { 1 },
+            suspend { 2 },
+            suspend { 3 }
+          ).flatTraverse {
+            listOf(it * 2)
+          }
+          result
+        }
+      } shouldBe listOf(2, 4, 6)
+    }
+
+    "List.traverse syntax" {
+      fxTest {
+        fx {
+          listOf(
+            suspend { 1 },
+            suspend { 2 },
+            suspend { 3 }
+          ).traverse(::effectIdentity)
+        }
+      } shouldBe listOf(1, 2, 3)
+    }
+
+    "List.sequence syntax" {
+      fxTest {
+        fx {
+          listOf(
+            suspend { 1 },
+            suspend { 2 },
+            suspend { 3 }
+          ).sequence()
+        }
+      } shouldBe listOf(1, 2, 3)
+    }
+
+    "List.parTraverse syntax" {
+      val main = Thread.currentThread().name
+      fxTest {
+        fx {
+          listOf(
+            suspend { Thread.currentThread().name },
+            suspend { Thread.currentThread().name },
+            suspend { Thread.currentThread().name }
+          ).parTraverse(
+            Dispatchers.Default,
+            ::effectIdentity
+          ).any {
+            it == main
+          }
+        }
+      } shouldBe false
+    }
+
+    "fx can turn effects into pure kinded values" {
+      suspend fun sideEffect(): String {
+        println("Boom!")
+        return ""
+      }
+      fxTest {
+        fx {
+          val (result) = f { sideEffect() }
+          result
+        }
+      } shouldBe ""
+    }
+
+  }
+}
+
+fun <A> fxTest(f: () -> IO<A>): A =
```

Comment:
This is the `unsafeRunBlocking` alias I told you about the other day may be useful for cases like this :)

Reply:
Yes, but this goes agains't the idea that effect can only run on `unsafe` via explicit imports. Since the goal of the Fx system is to provide higher compile time guarantees when dealing with effects we aim for a single way to run effects. Users can create such functions and name them `runThisIsFineNothingWillHappen()` if they want to but we shouldn't provide alias for unsafe runs in the environment. The Arrow Fx system treats the environment as `pure` an `suspend` as impure therefore running should be suspended and only materialized in one place by the user.